main/lsp/requests/code_action.cc

```diff
@@ -29,8 +29,229 @@ vector<unique_ptr<TextDocumentEdit>> getEdits(const LSPConfiguration &config, co
+        const auto newName = absl::StrCat(moduleName, i);
+        if (nameNotExists(newName)) {
+            return newName;
+        }
+    }
+
+    // bail, if we haven't found untaken name in 42 tries
+    return nullopt;
+}
+
+class MethodCallSiteRenamer : public AbstractRenamer {
+public:
+    MethodCallSiteRenamer(const core::GlobalState &gs, const LSPConfiguration &config, const string oldName,
+                          const string newName)
+        : AbstractRenamer(gs, config, oldName, newName) {
+        const vector<string> invalidNames = {"initialize", "call"};
+        for (auto name : invalidNames) {
+            if (oldName == name) {
+                invalid = true;
+                error = fmt::format("The `{}` method cannot be extracted to a module.", oldName);
+                return;
+            }
+        }
+        // block any method not starting with /[a-zA-Z0-9_]+/. This blocks operator overloads.
+        if (!isalnum(oldName[0]) && oldName[0] != '_') {
+            error = fmt::format("The `{}` method cannot be extracted to a module.", oldName);
+            invalid = true;
+        }
+    }
+
+    ~MethodCallSiteRenamer() {}
+    void rename(unique_ptr<core::lsp::QueryResponse> &response) override {
+        if (invalid) {
+            return;
+        }
+        auto loc = response->getLoc();
+
+        // If we're renaming the exact same place twice, silently ignore it. We reach this condition when we find the
+        // same method send through multiple definitions (e.g. in the case of union types)
+        auto it = edits.find(loc);
+        if (it != edits.end()) {
+            return;
+        }
+
+        auto source = loc.source(gs);
+        if (!source.has_value()) {
+            return;
+        }
+        string newsrc;
+        if (auto sendResp = response->isSend()) {
+            edits[sendResp->receiverLoc] = newName;
+        }
+    }
+    void addSymbol(const core::SymbolRef symbol) override {
+        if (symbol.isMethod()) {
+            addSubclassRelatedMethods(gs, symbol.asMethodRef(), getQueue());
+        } else {
+            getQueue()->tryEnqueue(symbol);
+        }
+    }
+}; // CallSiteRenamer
+
+vector<unique_ptr<TextEdit>> moveMethod(const LSPConfiguration &config, const core::GlobalState &gs,
+                                        const core::lsp::DefinitionResponse *definition,
+                                        LSPTypecheckerInterface &typechecker, string_view newModuleName) {
+    auto moduleStart = fmt::format("module {}\n  extend T::Sig\n  ", newModuleName);
+    auto moduleEnd = "\nend";
+
+    auto fref = definition->termLoc.file();
+    auto &file = fref.data(gs);
+
+    auto trees = typechecker.getResolved({fref});
+    auto &rootTree = trees[0].tree;
+    auto beginLoc = core::Loc::offset2Pos(fref.data(gs), rootTree.loc().beginPos());
+
+    auto pos2Loc = [&](const auto &pos) {
+        auto loc = core::Loc::offset2Pos(fref.data(gs), pos);
```

Comment:
Note that in general, I'm almost never a fan of capturing `gs` in a lambda function like this. When we're writing reusable functions that work over `gs`, I very much prefer that we make them top-level helper functions.